gcc/rust/typecheck/rust-hir-type-check-pattern.cc

```diff
@@ -262,16 +262,35 @@ TypeCheckPattern::visit (HIR::TuplePattern &pattern)
 	  = *static_cast<HIR::TuplePatternItemsMultiple *> (
 	    pattern.get_items ().get ());
 
+	size_t items = ref.get_patterns ().size ();
+
+	if (parent->get_kind () != TyTy::TUPLE)
+	  {
+	    rust_error_at (pattern.get_locus (), "expected %s, found tuple",
+			   parent->as_string ().c_str ());
+	    break;
+	  }
+
+	TyTy::TupleType &par = *static_cast<TyTy::TupleType *> (parent);
+	if (ref.get_patterns ().size () != par.get_fields ().size ())
+	  {
+	    size_t fields_cnt = par.get_fields ().size ();
+	    size_t patterns_cnt = ref.get_patterns ().size ();
+	    rust_error_at (pattern.get_locus (),
+			   "expected a tuple with %lu %s, found one "
+			   "with %lu %s",
+			   (unsigned long) fields_cnt,
+			   fields_cnt == 1 ? "element" : "elements",
+			   (unsigned long) patterns_cnt,
+			   patterns_cnt == 1 ? "element" : "elements");
```

Comment:
For an error as well done as this one is, can you extract it our as a separate function so that it becomes:

```emit_pattern_size_error (expected_field_count, got_field_count)```

We have been starting to extract out error diagnostics like this as separate functions so that they can be improved over time. One thing you might like to try is:

```
RichLocation r(pattern.get_locus())
r.add_Range(mappings->lookup_location(parent->get_Ref()));
rust_error_at(r, "error-message")
```